memory_allocator.c

```diff
@@ -0,0 +1,258 @@
+#include <common.h>
+#include <stdint.h>
+#include <stddef.h>
+#include "queue.h"
+#include "libkern.h"
+#include "include/memory_allocator.h"
+
+/*
+  TODO:
+  - use the mp_next field of malloc_pool
+*/
+
+#define MAGIC 0xC0DECAFE
+#define ALIGNMENT 8
+
+TAILQ_HEAD(mb_list, mem_block);
+TAILQ_HEAD(mp_arena, mem_arena);
+
+static inline int32_t abs(int32_t x)
+{
+  if (x < 0)
+    return -x;
+  return x;
+}
+
+typedef struct mem_block {
+  uint32_t mb_magic; /* if overwritten report a memory corruption error */
+  int32_t mb_size; /* size < 0 => free, size > 0 => alloc'd */
+  TAILQ_ENTRY(mem_block) mb_list;
```

Comment:
The list of used blocks might be useful for debugging, because we keep track of every block. Of course it might be more of a convenience thing, because we can recreate the list of free blocks: everything that isn't a free block is a used block.

That's a clever idea, but from my understanding, that union won't save us any space, because uint64_t mb_data[0] doesn't use any memory: it just might add padding to the end, so that the structure's size is a multiple of 8.
Without saving any space, I'd personally leave the used blocks be, because they use only about 16 bytes in every `mem_arena_t`.